File: monitor.py
#!/usr/bin/env python3
"""
Monitor de ONUs desautorizadas — roda separado do servidor Node.js
Uso: python3 monitor.py
Verifica a cada 5 minutos e salva log em server/monitor_log.json
"""
import json, time, os, urllib.request, urllib.error
from datetime import datetime

LOG_FILE = os.path.join(os.path.dirname(__file__), 'server', 'monitor_log.json')
API_URL  = 'http://localhost:3001/api/alertas'
INTERVAL = 5 * 60  # 5 minutos

def load_log():
    try:
        with open(LOG_FILE, 'r', encoding='utf-8') as f:
            return json.load(f)
    except:
        return []

def save_log(log):
    with open(LOG_FILE, 'w', encoding='utf-8') as f:
        json.dump(log[:200], f, ensure_ascii=False, indent=2)

def check():
    ts = datetime.now().isoformat()
    try:
        with urllib.request.urlopen(API_URL, timeout=10) as r:
            data = json.loads(r.read())

        counts = data.get('counts', {})
        desaut = counts.get('desautorizadas', 0)
        sem_st = counts.get('semStatus', 0)
        offline = counts.get('offline', 0)

        entry = {
            'ts': ts,
            'desautorizadas': desaut,
            'semStatus': sem_st,
            'offline': offline,
            'alertas': [
                {'login': r.get('Login'), 'mac': r.get('MAC/Serial'),
                 'nome': r.get('nome_formatado'), 'pon': r.get('PON ID')}
                for r in data.get('desautorizadas', [])
            ]
        }

        log = load_log()
        log.insert(0, entry)
        save_log(log)

        if desaut > 0 or sem_st > 0 or offline > 0:
            print(f'[{ts}] ⚠  Desautorizadas: {desaut} | Sem status: {sem_st} | Offline: {offline}')
            for a in entry['alertas']:
                print(f'   → {a["nome"]} | Login: {a["login"]} | PON: {a["pon"]}')
        else:
            print(f'[{ts}] ✓  Rede OK — sem alertas')

    except urllib.error.URLError as e:
        print(f'[{ts}] ✗  Servidor offline ou inacessível: {e.reason}')
    except Exception as e:
        print(f'[{ts}] ✗  Erro: {e}')

if __name__ == '__main__':
    print('=' * 55)
    print(' Monitor ISP — ONUs desautorizadas')
    print(f' Verificando a cada {INTERVAL//60} minutos')
    print(f' API: {API_URL}')
    print('=' * 55)
    while True:
        check()
        time.sleep(INTERVAL)

File: server/monitor.js
/**
 * Monitor de ONUs desautorizadas — roda a cada 5 minutos
 * Para usar: require('./monitor') no index.js
 * Loga no console e salva em monitor_log.json
 */
const fs   = require('fs');
const path = require('path');

const LOG_FILE = path.join(__dirname, 'monitor_log.json');
let log = [];
function loadLog() {
  if (fs.existsSync(LOG_FILE))
    try { log = JSON.parse(fs.readFileSync(LOG_FILE,'utf8')); } catch(e) {}
}
loadLog();

function saveLog(entry) {
  log.unshift(entry);
  if (log.length > 200) log = log.slice(0, 200);
  fs.writeFileSync(LOG_FILE, JSON.stringify(log, null, 2));
}

function startMonitor(getOnus) {
  const CHECK_MS = 5 * 60 * 1000; // 5 min

  function check() {
    const onus   = getOnus();
    const desaut = onus.filter(r => r['Status ONU'] === 'Desautorizada');
    const sem    = onus.filter(r => r['Status ONU'] === 'Sem status');
    const entry  = {
      ts:            new Date().toISOString(),
      total:         onus.length,
      desautorizadas: desaut.length,
      semStatus:     sem.length,
      alertas:       desaut.map(r => ({
        login: r.Login, mac: r['MAC/Serial'], pon: r['PON ID'], nome: r['Nome Cliente']
      })),
    };
    if (desaut.length > 0 || sem.length > 0) {
      console.warn(`[Monitor ${entry.ts}] ⚠ ${desaut.length} desautorizadas | ${sem.length} sem status`);
    }
    saveLog(entry);
  }

  // Run immediately, then every 5min
  check();
  setInterval(check, CHECK_MS);
  console.log('📡 Monitor de desautorizadas ativo (a cada 5 min)');
}

function getLog() { return log; }

module.exports = { startMonitor, getLog };
